File: cpp/040_CombinationSumII.cpp
class Solution {
public:
    vector<vector<int>> combinationSum2(vector<int>& candidates, int target) {
        std::sort(candidates.begin(), candidates.end());
        vector<vector<int>> ret;
        vector<int> path;
        backtrace(ret,path,candidates,target,0);
        return ret;
    }
private:
    void backtrace(vector<vector<int>>& ret, vector<int>& path, vector<int>& candidates, int target, int start){
        if(target == 0){
            ret.push_back(path);
            return;
        }
        if(target < 0) return;
        for(int i = start; i < candidates.size(); ++i){
            if(i > start && candidates[i] == candidates[i-1]) continue;
            path.push_back(candidates[i]);
            backtrace(ret,path,candidates,target-candidates[i],i+1);
            path.pop_back();
        }
    }
};

File: js/040_CombinationSumII.js
/**
 * @param {number[]} candidates
 * @param {number} target
 * @return {number[][]}
 */
var combinationSum2 = function(candidates, target) {
    candidates.sort( function(a, b) { return a-b; });
    var ret = [];
    backtrace(ret,[],candidates,target,0);
    return ret;
};

var backtrace = function(ret,path,candidates,target,start){
    if(target === 0){
        ret.push(path.concat());
        return;
    }
    if(target < 0) return;
    for(var i = start; i < candidates.length; ++i){
        if(i > start && candidates[i] === candidates[i-1]) continue;
        path.push(candidates[i]);
        backtrace(ret,path,candidates,target-candidates[i],i+1);
        path.pop();
    }
};